### PR TITLE
Remove unnecessary DMatrix methods

### PR DESCRIPTION
**Pull request by RAMitchell**
Two methods are removed that do not provide broadly useful functionality:
DMatrix::GetColumnDensity
DMatrix::SingleColumnBlock

Column density is only needed by the exact algorithm, so I compute it in the updater itself. Single column block can be established by other means.

### CODE REVIEW

File: src/linear/updater_gpu_coordinate.cu [hunk 1]
```diff
@@ -58,10 +58,15 @@ class GPUCoordinateUpdater : public LinearUpdater {  // NOLINT
 
     num_row_ = static_cast<size_t>(p_fmat->Info().num_row_);
 
-    CHECK(p_fmat->SingleColBlock());
-    SparsePage const& batch = *(p_fmat->GetBatches<CSCPage>().begin());
+    auto batch_iter = p_fmat->GetBatches<CSCPage>().begin();
```
Comment (trivialfis) [inline src/linear/updater_gpu_coordinate.cu]: I'm not sure why you want to remove this method and then created two different approaches to obtain the check.  One in gbm with dynamic cast another one here.

Comment (RAMitchell) [reply src/linear/updater_gpu_coordinate.cu]: I can use the dynamic cast approach here for consistency.

Comment (trivialfis) [reply src/linear/updater_gpu_coordinate.cu]: It seems that single method is more simple, we don't need to include extra header here, it breaks the abstraction.

Comment (RAMitchell) [reply src/linear/updater_gpu_coordinate.cu]: I reverted the change. Note that the abstraction is broken just by having a method in a base class that differentiates behaviour based on the subclass.

File: src/gbm/gbtree.cc [hunk 2]
```diff
@@ -27,6 +27,7 @@
 #include "../common/common.h"
 #include "../common/random.h"
 #include "../common/timer.h"
+#include "../data/sparse_page_dmatrix.h"
```
Comment (trivialfis) [inline src/gbm/gbtree.cc]: Revert this also.

File: src/linear/updater_gpu_coordinate.cu [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #include "../common/device_helpers.cuh"
 #include "../common/timer.h"
 #include "./param.h"
+#include "../data/sparse_page_dmatrix.h"
```
Comment (trivialfis) [inline src/linear/updater_gpu_coordinate.cu]: And this.

**Pull request merged**